src/theory/uf/proof_equality_engine.h

```diff
@@ -0,0 +1,340 @@
+   * Notice that noExplain is a subset of exp.
+   *
+   * The proof for conc follows from exp ^ expn by proof rule with the given
+   * id and arguments.
+   *
+   * This call corresponds to a conflict if conc is false and expn is empty.
+   *
+   * This returns the TrustNode corresponding to the formula corresonding to
+   * the call to this method [*], for which a proof can be provided by this
+   * generator in the remainder of the user context.
+   *
+   * [*]
+   * a. If this call does not correspond to a conflict, then this formula is:
+   *   ( ^_{e in exp} <explain>(e) ^ expn ) => conc
+   * where <explain>(e) is a conjunction of literals L1 ^ ... ^ Ln such that
+   * L1 ^ ... ^ Ln entail e, and each Li was passed as an explanation to a
+   * call to assertFact in the current SAT context. This explanation method
+   * always succeeds, provided that e is a literal that currently holds in
+   * the equality engine of this class. Notice that if the antecedant is empty,
+   * the formula above is assumed to be conc itself. The above formula is
+   * intended to be valid in Theory that owns this class.
+   * b. If this call is a conflict, then this formula is:
+   *   ^_{e in exp} <explain>(e)
+   * The formula can be queried via TrustNode::getProven in the standard way.
+   */
+  TrustNode assertLemma(Node conc,
+                        PfRule id,
+                        const std::vector<Node>& exp,
+                        const std::vector<Node>& noExplain,
+                        const std::vector<Node>& args);
+  /** Multi-step version */
+  TrustNode assertLemma(Node conc,
+                        const std::vector<Node>& exp,
+                        const std::vector<Node>& noExplain,
+                        ProofStepBuffer& psb);
+  /** Generator version, where pg has a proof of conc */
+  TrustNode assertLemma(Node conc,
+                        const std::vector<Node>& exp,
+                        const std::vector<Node>& noExplain,
+                        ProofGenerator* pg);
+  //-------------------------- explain
+  /**
+   * Explain literal conc. This calls the appropriate methods in the underlying
+   * equality engine of this class to construct the explanation of why conc
+   * currently holds.
+   *
+   * It returns a trust node of kind TrustNodeKind::PROP_EXP whose node
+   * is the explanation of conc (a conjunction of literals that implies it).
+   * The proof that can be proven by this generator is then (=> exp conc), see
+   * TrustNode::getPropExpProven(conc,exp);
+   *
+   * @param conc The conclusion to explain
+   * @return The trust node indicating the explanation of conc and the generator
+   * (this class) that can prove the implication.
+   */
+  TrustNode explain(Node conc);
+
+ protected:
+  /** Add proof step */
+  bool addProofStep(Node lit,
+                    PfRule id,
+                    const std::vector<Node>& exp,
+                    const std::vector<Node>& args);
+  /** Assert internal */
```

Comment:
say what this method does, in particular that it is responsible for adding stuff to `d_keep`

Reply:
Actually this one is unused, i've deleted.